### ISSUE TITLE
Enable pre/post/same destructiveChanges using keywords in commit messages

### ISSUE THREAD

**Issue by AllanOricil** [scolladon/sfdx-git-delta#583]
### Is your proposal related to a problem?

I can't determine if the changes generated by the plugin should be deleted before, after or in the same deployment job. I think that the solution for this problem is to include "reserved keywords" in git messages that sgd could use. Why version this information? Because I could replay the branch history to determine the order metadata changes happened, including deletions.

### Describe a solution you propose

Given this set of commits
`````bash
23echsd      [sgd-delete-before] commit message 1
84jdhe2       commit message 2
528jded       [sgd-delete-after] commit messages 3
`````

And considering HEAD is pointing to 528jded,
when running sgd to get the delta from `23echsd` to HEAD, the result should generate 4 manifests

preDestructiveChanges.xml
postDestructiveChanges.xml
destructiveChanges.xml
package.xml

### Describe alternatives you've considered

Currently there is no alternative. I would have to  read commits again and run sgd multiple times between every commit if I create my own solution.

### Additional context

N/A

**Comment by AllanOricil** [comment]
Or maybe there is another way you see people using. I imagine that they use PR comments, and I don't think this is good because the information of the order things get added and removed should also be versioned.

**Comment by AllanOricil** [comment]
The could also exist a keyword to mark deployment job boundaries. This way we can create some sort of deployment in multiple steps using commits:


````bash
23echsd     [sgd-delete-before] commit message 1
84jdhe2       commit message 2
528jded      [sgd-deploy-boundary] [sgd-delete-after] commit messages 3
63ecdsd     [sgd-delete-before] commit message 1
94jwhe2       commit message 2
a28jded      [sgd-deploy-boundary] [sgd-delete-after] commit messages 3
````

When multiple `[sgd-deploy-boundary]` are found, sgd will generate multiple folders to separate the manifests.

````bash
deploy_boundary_1
     preDestructiveChanges.xml
     postDestructiveChanges.xml
     destructiveChanges.xml
     package.xml
deploy_boundary_2
     preDestructiveChanges.xml
     postDestructiveChanges.xml
     destructiveChanges.xml 
    package.xml
````


The json output would have information that can easily help us to determine how many deployments sgd created

**Comment by scolladon** [comment]
Hi @AllanOricil !

Thanks for raising this enhancement request and thanks for contributing in making this project better!

Could you create another Enhancement Request for the "deployment job boundaries" concept?
This way we will be able to work on them separately.

I'm very interested in both concepts.
I think it can be very useful for some very sharp DevOps situations.
I cannot imagine easily what those DevOps workflow could be and if there is no simpler design to address the issue it addresses, but I'm interested in the concepts :) 
I'll think about this in the next following days, what use cases it enables, what problems it solves, is there other suitable ways to approach this, and see what would be a suitable software design to easily ship this.

For the destructive document part, why do you think we should distinguish `before` from `same` ? To me a `destructiveChanges.xml` is [always applied before deployment](https://developer.salesforce.com/docs/atlas.en-us.daas.meta/daas/daas_destructive_changes.htm#:~:text=deletions%20are%20processed%20before). So I think we could consider `after` tag only. If the commit is not tagged then it should be added in the `destructiveChangesPre.xml` (nominal use case).
In the end we would have the `package.xml`, the `destructiveChangesPre.xml` (instead of the `destructiveChanges.xml`) and the `destructiveChangesPost.xml`.
Do you have any specific use case using 3 destructiveChanges? What is your idea behind keeping the `destructiveChanges.xml`?

**Comment by AllanOricil** [comment]
> For the destructive document part, why do you think we should distinguish `before` from `same` ? To me a

Yes, I thought that destructiveChanges.xml would run in the same deployment job. But according to the docs, it runs before.



> So I think we could consider `after` tag only. If the commit is not tagged then it should be added in the `destructiveChangesPre.xml` (nominal use case).

Yes, there is no need for using other keywords besides `after`

**Comment by scolladon** [comment]
### What I understand from the need

The need is to have a way to deal with `pre` and `post`  destructive content, driven by the repository.
By default (when not splitted), content in the `destructiveChanges.xml` is executed [before](https://developer.salesforce.com/docs/atlas.en-us.daas.meta/daas/daas_destructive_changes.htm#:~:text=deletions%20are%20processed%20before).
`sfdx-git-delta` should be able to deduce where to put deleted element (in `pre` or in `post`) from reading the git repository.

### What I understand from the solution proposed

Any deletion should be added in the `destructiveChangesPre.xml` (it works the same way as if it was in the `destructiveChanges.xml`)
When a deletion is in a commit flagged with a keyword (`[sgd-delete-before]` for instance) in the commit message (or git note) then the deletion is added in the `destructiveChangesPost.xml`

#### Edge cases discussion

I can see the value in CI when evaluating a PR.
In this context we can have multiple commits against the target branch. 
It is easy to then only validate the delta with the target branch and the diff will allow to move the deleted element in `pre` or `post` `destructiveChanges.xml`.
But, when this PR is merged it will be one commit in the target branch (when merged with squash and merge)... If we want to do the same deployment into a QA environment for example when it is merged in the target branch then it will not be possible to have this granularity and to know what to scope to put in the `destructiveChangesPost.xml`.
So I think this solution would not work for this kind of workflow and would force the user in a specific workflow to benefit from this capability.

### Alternative approach

I don't have one so far... 😅 
I really like reading the commit message, it is easy, even if the actual architecture is not suitable for this kind of work I think it could be done in post processing for example.
But I don't know how we could have a solution to detect when a change needs to be put into the `destructiveChangesPost.xml` based on git content, for any usage...

Let's think about it further, maybe @AllanOricil you'll get a good idea on how to detect what metadata to put in `destructiveChangesPost.xml` whatever the kind of commit (squash / merge / range)

**Comment by AllanOricil** [comment]
I thought about a different approach. Tell me what do you think about it.


There could exist a file called `.sgdconfig.yml` in the root of an sfdx project. It has the following pattern:


````yml
...

delete-configuration:
    before:
       # I need to delete this class before adding that other one
       - package: my-package-a 
         type: ApexClass
         name: Foo.cls

    after:
       - package: my-package-b
         type: ApexClass
         name: Bar.cls
````

There could also exist a schema defining the structure of this file so that it can be validated while writting it in vscode, or any other text editor that has support for schema validation. Having a schema validating the structure of the file also ensures developers won't commit a file that has mistakes.


`sgd` must try to fetch this file from the git ref passed to the `--to` flag.

**Comment by AllanOricil** [comment]
What if `sgd` exposes a new flag that accepts a list of metadata paths that has to be deleted using `destructiveChangesPost.xml`.  CICD pipelines can fetch this information from PR descriptions or commit messages, and pass it to `sgd` using the format it is expecting. 

This flag could be called `--delete-after`:


`sgd ... --delete-after ./package/main/default/classes/Foo.cls --delete-after ./package/main/default/classes/Bar.cls ...`


`sgd` would need to take care about these things:

1. Verify if the given path is really deleted. If not, display an error and exit 1.
2. generate both destructive changes manifests, but If all deleted files paths were passed to `--delete-after`, skip the creation of the `destructiveChangesPre.xml`



There could also exist a flag to accept a path to a file that has a json containing an array of metadata paths. This is useful for huge releases with lots of files that can only be deleted after

````json
[
  "./packageA/main/default/classes/Foo.cls",
  "./packageB/main/default/classes/Bar.cls"
]
````

**Comment by AllanOricil** [comment]
It is the same as if we were using Salesforce's UI to configure which metadata items have to be deleted afterwards

https://success.flosum.com/s/article/Destructive-Changes

The UI in this case would be Gitlab/Github/Bitbucket PR description/comment. The input would be parsed to a format that `sgd` wants.

**Comment by scolladon** [comment]
Yep I agree @AllanOricil, It is the same as we were using a graphical UI instead of a CLI [here](https://github.com/scolladon/sfdx-git-delta/issues/583#issuecomment-1539186539).

I think point 4 [here](https://github.com/scolladon/sfdx-git-delta/issues/583#issuecomment-1536939962) is debatable (just saying 🙂)

I like the `--delete-after`, it seems to address the issue of the squash and merge commit making disappear the information when the information is read from the commit message but the UX is does not look very convenient yet.
But it looks like it could be done by versioning a `destructiveChangesPost.xml` and remove its content from the `destructiveChanges(Pre).xml` via a simple command / tool (like `sed`). 
The advantage for the parameter would be it does not require to empty the `destructiveChangesPost.xml` after a release.

I'll try to think on that in the coming days, and let you know if I have an idea.
Feel free to improve if you have inspiration 🙂

**Comment by AllanOricil** [comment]
Is there some sort of tool that creates a release plan based on git changes + metadata dependency api?

**Comment by AllanOricil** [comment]
@scolladon 


I thought again about storing this information on git commits and I think I found a way. See if it makes sense:


Branching model: git flow


1. feature/bugfix branches commits stores the list of metadata that has to be deleted after in a format that sgd can parse

````bash
...

>>sgd
path/to/metadata/1
path/to/metadata/2
<<sgd
````

2. sgd always look for this information from HEAD. If it does not find it, then it considers that all deletions happen before.

3. feature/bugfix branches are always rebased and must not contain any merge commits.

4. feature/bugfix branches are always squashed and merged to develop. When doing so, the developer has to edit the commit message with the latest list of metadata files that have to be deleted after. There can't exist 2   sgd delimiters (>>sgd <<sgd) in the commit message.

5. release branches are created from develop and merged to main with a merge commit so that the commit history remains unchanged

6. When pushing the release branch to remote, my workflow would iterate over the list of feature/bugfix commits and deploy each one of them in sequence, scheduling multiple deployment jobs. Each deployment job represents a "change set" with files that have to be deleted before and after.

**Comment by Bartheleway** [comment]
To join this little debate as I am also facing this question. I would suggest to have a "deployOptions" folder which contains a folder with branch name (often containing issue/feature number) so that the developer can decide if deletion has to happen in pre or post. Git delta will then just merge all new files (xml) in this folder which will provide correct behavior for pre / post deletion.

Later on Git maintainer can archive/delete these files when he decide they are not necessary anymore.

**Comment by scolladon** [comment]
I think this idea should be done outside of the plugin @Bartheleway.
It don't see it very hard, it could be done easily with a merge [package.xml plugin](https://sfdx-hardis.cloudity.com/hardis/package/mergexml/) with two commands, one for the pre and another one for the post.
I hope it helps !
